homeassistant/components/laundrify/sensor.py

```diff
@@ -0,0 +1,108 @@
+"""Platform for sensor integration."""
+
+import logging
+
+from laundrify_aio import LaundrifyDevice
+from laundrify_aio.exceptions import LaundrifyDeviceException
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import UnitOfEnergy, UnitOfPower
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN
+from .coordinator import LaundrifyUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, config: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Add power sensor for passed config_entry in HA."""
+
+    coordinator: LaundrifyUpdateCoordinator = hass.data[DOMAIN][config.entry_id][
+        "coordinator"
+    ]
+
+    sensor_entities: list[LaundrifyPowerSensor | LaundrifyEnergySensor] = []
+    for device in coordinator.data.values():
+        sensor_entities.append(LaundrifyPowerSensor(device))
+        sensor_entities.append(LaundrifyEnergySensor(coordinator, device))
+    if sensor_entities:
+        async_add_entities(sensor_entities)
+
+
+class LaundrifyPowerSensor(SensorEntity):
+    """Representation of a Power sensor."""
+
+    _attr_device_class = SensorDeviceClass.POWER
+    _attr_native_unit_of_measurement = UnitOfPower.WATT
+    _attr_state_class = SensorStateClass.MEASUREMENT
+    _attr_suggested_display_precision = 0
+
+    _attr_has_entity_name = True
+    _attr_translation_key = "power_sensor"
+
+    def __init__(self, device: LaundrifyDevice) -> None:
+        """Initialize the sensor."""
+        super().__init__()
+        self._device = device
+        self._attr_device_info = {"identifiers": {(DOMAIN, device.id)}}
+        self._attr_unique_id = f"{device.id}_power"
+
+    async def async_update(self) -> None:
+        """Fetch latest power measurement from the device."""
+        try:
+            power = await self._device.get_power()
```

Comment:
This integration uses a coordinator, shouldn't that be used instead?

Reply:
I chose to use a non-coordinated entity for the `PowerSensor` since the latest power measurements are polled locally from each individual power plug.

Opposed to that, the latest updates for the `EnergySensor` and `BinarySensor` are fetched from the backend in a single call (using the `LaundrifyUpdateCoordinator`).

Reply:
My understanding of a coordinator was to group data polling for several entities/devices to a single, coordinated API call.

Using a coordinator per device sounds a bit counter-intuitive and needlessly complex to me. As per the [developer docs](https://developers.home-assistant.io/docs/integration_fetching_data#separate-polling-for-each-individual-entity):

> If you can map exactly one device endpoint to a single entity, you can fetch the data for this entity inside the `update()`/`async_update()` methods. 

Let me know if I miss something.

Reply:
I think there are several ways to look at how you manage state in an integration. For example if every entity has its own endpoint, it would make sense to use `async_update` everywhere. But if every entity is using a centralized method to fetch the state, except for one, and there is no reason why we can't merge those (think rate limiting, different polling intervals) it would also make sense to just get all data in the single fetch. The added benefit is that your unavailability handling is smoother (because polling could happen at different intervals).

And as added benefit, I think this code could be a lot cleaner if it were just 2 sensor entities exposing the state from the coordinator

Reply:
Just to make sure we have a common understanding of what the two entities (`EnergySensor` and `PowerSensor`) are doing:
 - `EnergySensor` fetches the total energy of a device _from the backend_. A coordinator is used here, because we are calling the backend and can fetch the data in a single call.
 - `PowerSensor` fetches the latest power measurement _from the physical device directly_. Since each physical device has another IP address, we cannot consolidate these calls to a single call. This is why it uses `async_update()` instead of the coordinator.